File: labs/M08-conversation-management/solution/auto_summarize.py
"""
M08 Lab - Step 3: Auto-Summarizing Conversation Manager (Solution)
===================================================================
Complete solution: an AutoSummarizeManager that compresses old messages
into a summary when the conversation hits 80% of its token budget.

Usage:
    python auto_summarize.py
"""

import json
from dotenv import load_dotenv

load_dotenv()

import anthropic

client = anthropic.Anthropic()
MODEL = "claude-sonnet-4-20250514"

SYSTEM_PROMPT = (
    "You are a UCC filing research assistant. Help users understand "
    "UCC filings, lien risks, and secured transactions. Provide clear, "
    "concise answers. When referencing prior conversation, demonstrate "
    "you remember the context."
)


# =============================================================================
# OBSERVATION HELPERS
# =============================================================================

def observe(label: str, message: str) -> None:
    """Print a labeled observation line."""
    print(f"\n{'=' * 60}")
    print(f"[{label}] {message}")
    print(f"{'=' * 60}")


def observe_tokens(token_count: int, max_tokens: int, message_count: int) -> None:
    """Log token usage against budget."""
    pct = (token_count / max_tokens * 100) if max_tokens > 0 else 0
    threshold_marker = " *** ABOVE 80% THRESHOLD ***" if pct >= 80 else ""
    print(f"[TOKENS] ~{token_count} / {max_tokens} ({pct:.0f}%){threshold_marker}")
    print(f"[HISTORY] {message_count} messages in history")


def observe_summarize(num_messages: int, tokens_before: int, tokens_after: int) -> None:
    """Log summarization event."""
    print(f"[SUMMARIZE] Compressed {num_messages} messages into summary "
          f"({tokens_before} tokens -> {tokens_after} tokens)")


# =============================================================================
# SOLUTION: AutoSummarizeManager
# =============================================================================

class AutoSummarizeManager:
    """
    Manages conversation history with automatic summarization.

    When the conversation reaches 80% of the token budget, the oldest
    messages are sent to Claude for summarization. The summary replaces
    those messages, dramatically reducing token usage while preserving
    context.
    """

    def __init__(
        self,
        system_prompt: str = SYSTEM_PROMPT,
        max_tokens: int = 2048,
        threshold: float = 0.8,
    ):
        """Initialize the auto-summarize manager."""
        self.system_prompt = system_prompt
        self.max_tokens = max_tokens
        self.threshold = threshold
        self.summarize_count = 0
        # Step 1: Initialize empty messages list
        self.messages = []

    def _estimate_tokens(self, messages: list[dict]) -> int:
        """Estimate token count for a list of messages plus system prompt."""
        # Step 2: Estimate tokens
        return (len(self.system_prompt) + len(json.dumps(messages))) // 4

    def _should_summarize(self) -> bool:
        """Check if current token usage exceeds the threshold."""
        # Step 3: Check threshold
        return self._estimate_tokens(self.messages) >= self.max_tokens * self.threshold

    def _summarize_old_messages(self) -> None:
        """Compress old messages into a summary."""
        # Step 4: Calculate split point -- keep the last 1/3 of messages intact
        split_at = max(2, (len(self.messages) * 2) // 3)
        old_messages = self.messages[:split_at]
        recent_messages = self.messages[split_at:]

        # Step 5: Record tokens before summarization
        tokens_before = self._estimate_tokens(self.messages)

        # Step 6: Build summarization prompt from old messages
        conversation_text = ""
        for msg in old_messages:
            role = msg["role"].upper()
            content = msg["content"]
            conversation_text += f"{role}: {content}\n\n"

        summarize_prompt = (
            "Summarize this conversation so far in 2-3 sentences. Focus on "
            "the key topics discussed and any important facts established.\n\n"
            f"Conversation:\n{conversation_text}"
        )

        summary_response = client.messages.create(
            model=MODEL,
            max_tokens=256,
            messages=[{"role": "user", "content": summarize_prompt}],
        )

        # Step 7: Extract summary text
        summary_text = ""
        for block in summary_response.content:
            if hasattr(block, "text"):
                summary_text += block.text

        # Step 8: Build new messages list with summary + recent messages
        summary_message = {
            "role": "assistant",
            "content": f"[Summary of earlier conversation]: {summary_text}",
        }

        # Ensure role alternation: if first recent message is also "assistant",
        # insert a bridging user message
        new_messages = [summary_message]
        if recent_messages and recent_messages[0]["role"] == "assistant":
            new_messages.append({"role": "user", "content": "(continuing conversation)"})
        new_messages.extend(recent_messages)

        self.messages = new_messages

        # Step 9: Record tokens after and log
        tokens_after = self._estimate_tokens(self.messages)
        self.summarize_count += 1
        observe_summarize(len(old_messages), tokens_before, tokens_after)
        print(f'[SUMMARIZE] Summary: "{summary_text[:100]}..."')

    def add_user_message(self, text: str) -> None:
        """Add a user message to the conversation history."""
        # Step 10: Append user message
        self.messages.append({"role": "user", "content": text})

    def send(self) -> str:
        """Send the conversation to Claude and append the response."""
        # Step 11: Check if summarization is needed, then call API
        if self._should_summarize():
            self._summarize_old_messages()

        response = client.messages.create(
            model=MODEL,
            max_tokens=1024,
            system=self.system_prompt,
            messages=self.messages,
        )

        # Step 12: Extract text and append to history
        assistant_text = ""
        for block in response.content:
            if hasattr(block, "text"):
                assistant_text += block.text

        self.messages.append({"role": "assistant", "content": assistant_text})

        return assistant_text

    def get_history(self) -> list[dict]:
        """Return the current message history."""
        # Step 13: Return messages
        return self.messages

    def get_token_count(self) -> int:
        """Return estimated token count for current history."""
        return self._estimate_tokens(self.messages) if self.messages else 0


# =============================================================================
# MAIN
# =============================================================================

if __name__ == "__main__":
    print("=" * 60)
    print("M08 Lab - Step 3: Auto-Summarizing Conversation Manager (SOLUTION)")
    print("=" * 60)

    manager = AutoSummarizeManager(max_tokens=2048)
    threshold_tokens = int(manager.max_tokens * manager.threshold)
    print(f"Token budget: {manager.max_tokens} tokens "
          f"(summarize at 80% = {threshold_tokens} tokens)")

    test_questions = [
        "What is a UCC-1 filing?",
        "Who files a UCC-1?",
        "What is the purpose of perfecting a security interest?",
        "What collateral types can be covered by a UCC filing?",
        "What is a continuation statement?",
        "What is a UCC-3 amendment?",
        "How do I search for existing UCC filings?",
        "What are the risks of not filing a UCC-1?",
        "What is a purchase money security interest?",
        "How do UCC filings work in bankruptcy?",
        "What is a blanket lien?",
        "How do fixture filings work?",
        "What is a debtor-in-possession?",
        "What is the difference between attachment and perfection?",
        "Give me a final summary of everything we covered",
    ]

    for i, question in enumerate(test_questions, 1):
        print(f"\n--- Turn {i}/{len(test_questions)} ---")

        observe("USER", question)
        manager.add_user_message(question)

        response = manager.send()

        observe("ASSISTANT", response[:200] + "..." if len(response) > 200 else response)
        observe_tokens(
            manager.get_token_count(),
            manager.max_tokens,
            len(manager.get_history()),
        )

    print(f"\n{'=' * 60}")
    print(f"Final: {len(manager.get_history())} messages, "
          f"{manager.summarize_count} summarization events triggered")
    print(f"{'=' * 60}")

File: labs/M08-conversation-management/solution/auto_summarize.js
/**
 * M08 Lab - Step 3: Auto-Summarizing Conversation Manager (Solution)
 * ===================================================================
 * Complete solution: an AutoSummarizeManager that compresses old messages
 * into a summary when the conversation hits 80% of its token budget.
 *
 * Usage:
 *     node auto_summarize.js
 */

import Anthropic from "@anthropic-ai/sdk";
import "dotenv/config";

const client = new Anthropic();
const MODEL = "claude-sonnet-4-20250514";

const SYSTEM_PROMPT =
  "You are a UCC filing research assistant. Help users understand " +
  "UCC filings, lien risks, and secured transactions. Provide clear, " +
  "concise answers. When referencing prior conversation, demonstrate " +
  "you remember the context.";

// =============================================================================
// OBSERVATION HELPERS
// =============================================================================

function observe(label, message) {
  console.log(`\n${"=".repeat(60)}`);
  console.log(`[${label}] ${message}`);
  console.log("=".repeat(60));
}

function observeTokens(tokenCount, maxTokens, messageCount) {
  const pct = maxTokens > 0 ? (tokenCount / maxTokens) * 100 : 0;
  const marker = pct >= 80 ? " *** ABOVE 80% THRESHOLD ***" : "";
  console.log(
    `[TOKENS] ~${tokenCount} / ${maxTokens} (${Math.round(pct)}%)${marker}`
  );
  console.log(`[HISTORY] ${messageCount} messages in history`);
}

function observeSummarize(numMessages, tokensBefore, tokensAfter) {
  console.log(
    `[SUMMARIZE] Compressed ${numMessages} messages into summary ` +
      `(${tokensBefore} tokens -> ${tokensAfter} tokens)`
  );
}

// =============================================================================
// SOLUTION: AutoSummarizeManager
// =============================================================================

class AutoSummarizeManager {
  constructor(
    systemPrompt = SYSTEM_PROMPT,
    maxTokens = 2048,
    threshold = 0.8
  ) {
    this.systemPrompt = systemPrompt;
    this.maxTokens = maxTokens;
    this.threshold = threshold;
    this.summarizeCount = 0;
    // Step 1: Initialize empty messages array
    this.messages = [];
  }

  _estimateTokens(messages) {
    // Step 2: Estimate tokens
    return Math.floor(
      (this.systemPrompt.length + JSON.stringify(messages).length) / 4
    );
  }

  _shouldSummarize() {
    // Step 3: Check threshold
    return (
      this._estimateTokens(this.messages) >= this.maxTokens * this.threshold
    );
  }

  async _summarizeOldMessages() {
    // Step 4: Calculate split point -- keep the last 1/3 of messages intact
    const splitAt = Math.max(
      2,
      Math.floor((this.messages.length * 2) / 3)
    );
    const oldMessages = this.messages.slice(0, splitAt);
    const recentMessages = this.messages.slice(splitAt);

    // Step 5: Record tokens before summarization
    const tokensBefore = this._estimateTokens(this.messages);

    // Step 6: Build summarization prompt from old messages
    let conversationText = "";
    for (const msg of oldMessages) {
      const role = msg.role.toUpperCase();
      conversationText += `${role}: ${msg.content}\n\n`;
    }

    const summarizePrompt =
      "Summarize this conversation so far in 2-3 sentences. Focus on " +
      "the key topics discussed and any important facts established.\n\n" +
      `Conversation:\n${conversationText}`;

    const summaryResponse = await client.messages.create({
      model: MODEL,
      max_tokens: 256,
      messages: [{ role: "user", content: summarizePrompt }],
    });

    // Step 7: Extract summary text
    let summaryText = "";
    for (const block of summaryResponse.content) {
      if (block.text) {
        summaryText += block.text;
      }
    }

    // Step 8: Build new messages array with summary + recent messages
    const summaryMessage = {
      role: "assistant",
      content: `[Summary of earlier conversation]: ${summaryText}`,
    };

    // Ensure role alternation: if first recent message is also "assistant",
    // insert a bridging user message
    const newMessages = [summaryMessage];
    if (recentMessages.length > 0 && recentMessages[0].role === "assistant") {
      newMessages.push({
        role: "user",
        content: "(continuing conversation)",
      });
    }
    newMessages.push(...recentMessages);

    this.messages = newMessages;

    // Step 9: Record tokens after and log
    const tokensAfter = this._estimateTokens(this.messages);
    this.summarizeCount++;
    observeSummarize(oldMessages.length, tokensBefore, tokensAfter);
    console.log(`[SUMMARIZE] Summary: "${summaryText.slice(0, 100)}..."`);
  }

  addUserMessage(text) {
    // Step 10: Push user message
    this.messages.push({ role: "user", content: text });
  }

  async send() {
    // Step 11: Check if summarization is needed, then call API
    if (this._shouldSummarize()) {
      await this._summarizeOldMessages();
    }

    const response = await client.messages.create({
      model: MODEL,
      max_tokens: 1024,
      system: this.systemPrompt,
      messages: this.messages,
    });

    // Step 12: Extract text and push to history
    let assistantText = "";
    for (const block of response.content) {
      if (block.text) {
        assistantText += block.text;
      }
    }

    this.messages.push({ role: "assistant", content: assistantText });

    return assistantText;
  }

  getHistory() {
    // Step 13: Return messages
    return this.messages;
  }

  getTokenCount() {
    return this.messages.length > 0 ? this._estimateTokens(this.messages) : 0;
  }
}

// =============================================================================
// MAIN
// =============================================================================

console.log("=".repeat(60));
console.log(
  "M08 Lab - Step 3: Auto-Summarizing Conversation Manager (SOLUTION)"
);
console.log("=".repeat(60));

const manager = new AutoSummarizeManager(SYSTEM_PROMPT, 2048);
const thresholdTokens = Math.floor(manager.maxTokens * manager.threshold);
console.log(
  `Token budget: ${manager.maxTokens} tokens ` +
    `(summarize at 80% = ${thresholdTokens} tokens)`
);

const testQuestions = [
  "What is a UCC-1 filing?",
  "Who files a UCC-1?",
  "What is the purpose of perfecting a security interest?",
  "What collateral types can be covered by a UCC filing?",
  "What is a continuation statement?",
  "What is a UCC-3 amendment?",
  "How do I search for existing UCC filings?",
  "What are the risks of not filing a UCC-1?",
  "What is a purchase money security interest?",
  "How do UCC filings work in bankruptcy?",
  "What is a blanket lien?",
  "How do fixture filings work?",
  "What is a debtor-in-possession?",
  "What is the difference between attachment and perfection?",
  "Give me a final summary of everything we covered",
];

for (let i = 0; i < testQuestions.length; i++) {
  const question = testQuestions[i];
  console.log(`\n--- Turn ${i + 1}/${testQuestions.length} ---`);

  observe("USER", question);
  manager.addUserMessage(question);

  const response = await manager.send();

  observe(
    "ASSISTANT",
    response.length > 200 ? response.slice(0, 200) + "..." : response
  );
  observeTokens(
    manager.getTokenCount(),
    manager.maxTokens,
    manager.getHistory().length
  );
}

console.log(`\n${"=".repeat(60)}`);
console.log(
  `Final: ${manager.getHistory().length} messages, ` +
    `${manager.summarizeCount} summarization events triggered`
);
console.log("=".repeat(60));
